### PR TITLE
[Workspace] Upgrade rust and use new workspace inheritance features.

### PR DESCRIPTION
**Pull request by JoshLind**
### Description
This PR updates our rust version to 1.64.0. As an added benefit, it also makes use of the new workspace inheritance features to: (i) share workspace package properties across all crates (i.e., we avoid repeating all the same fields across all workspace crates); and (ii) pin dependencies across the entire workspace (i.e., we avoid depending on different versions of the same dependency in our crates and make upgrading dependencies super trivial!).

Note: Given this is somewhat of a big change, I've only changed a few dependencies to use this new functionality so the team can provide feedback/concerns. If we all agree, I can update all external dependencies to use this.

### Test Plan
Existing test infra.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4536)

### CODE REVIEW

**Review by davidiw — COMMENTED** [review]
very interesting... If we are gonna move to Rust 1.64, do we want to wait until after the mainnet cut?

**Comment by JoshLind** [comment]
> very interesting... If we are gonna move to Rust 1.64, do we want to wait until after the mainnet cut?

Yeah, I guess the trade-off here is between the risks involved with: (i) updating to a newer version of rust; vs (ii) continuing to allow multiple versions of the same dependencies across our crates.

Personally, I'm less worried about the rust upgrade... What I'll do for now is update the PR to be all-inclusive (so folks can see what it looks like) and we can decide if we want to land it pre vs. post cut.

**Comment by vgao1996** [comment]
I wonder if this can solve our issues with `move-deps`. `move-deps` was created so that we didn't have to add the move dependencies (referred to by git hashes) everywhere, but it created a nasty feature unification issue. I suppose this would allow us to declare the Move deps at workspace level (with the hashes) and use them in individual crates, potentially with different feature flags.

File: api/src/tests/transaction_vector_test.rs [hunk 1]
```diff
@@ -296,9 +296,8 @@ async fn test_script_payload() {
         for arg in args.as_array_mut().unwrap() {
             let arg_obj = arg.as_object_mut().unwrap();
 
-            match arg_obj.get_mut("U8Vector") {
-                Some(val) => *val = byte_array_to_hex(val),
-                None => {}
+            if let Some(val) = arg_obj.get_mut("U8Vector") {
+                *val = byte_array_to_hex(val)
```
Comment (Markuze) [inline api/src/tests/transaction_vector_test.rs:300]: Nit: does this belong to this PR?

**Review by Markuze — COMMENTED** [review]
Looks, good I'm happy for that this landed in 1.64.
Just a nit that you have a couple unrelated changes, and this is already a big change.

**Comment by bchocho** [comment]
Please add `rust-version` to avoid confusion after the upgrade as it will break compile in weird ways for anyone on an older version (learned the hard way trying to compile faucet)
https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field

**Comment by davidiw** [comment]
closing out stale pr, but lots of good stuff here to iterate on...

**Pull request closed without merging**